### ISSUE TITLE
Exception error raised when opening a CSV with large number of columns

### ISSUE THREAD

**Issue by jquartel** [OpenRefine/OpenRefine#6723]
I have some CSVs and TSVs which have hundreds of columns and which don't open in OpenRefine. I don't really mind that there is a limit on columns, but when it attempts to open such files (into the parsing page after selecting the file) a popup window is shown with a java exception. Here is the message in the popup I got when opening a TSV with 1176 columns:

java.lang.ArrayIndexOutOfBoundsException - Index 512 out of bounds for length 512
Parser Configuration: TsvParserSettings:
	Auto configuration enabled=true
	Auto-closing enabled=true
	Column reordering enabled=true
	Header extraction enabled=null
	Headers=null
	Ignore leading whitespaces=true
	Ignore trailing whitespaces=true
	Input buffer size=1048576
	Input reading on separate thread=true
	Length of content displayed on error=-1
	Line separator detection enabled=false
	Maximum number of characters per column=262144
	Maximum number of columns=512
	Null value=null
	Number of records to read=all
	Processor=none
	Restricting data in exceptions=false
	RowProcessor error handler=null
	Selected fields=none
	Skip bits as whitespace=true
	Skip empty lines=trueFormat configuration:
	TsvFormat:
		Comment character=#
		Escape character=\
		Line separator (normalized)=\n
		Line separator sequence=\n
Internal state when error was thrown: line=0, column=513, record=0, charIndex=18558

I think OR should, instead, throw up a short message stating the column limit (I'm guessing it's 512 based on the above error). It would be even better if OR could handle more columns.

I'm using OpenRefine 3.8.2 on Safari 17.5 on MacOS 14.5

**Comment by ostephens** [comment]
Suspect this was introduced when we switched to Univocity https://github.com/OpenRefine/OpenRefine/pull/6098. It would be nice if the number of columns parsed became an option that the user could set during import (at the moment we only expose the option to import a fixed number of rows)

**Comment by tfmorris** [comment]
@ostephens analysis is correct. I'm happy to make the max column count arbitrarily large since we didn't have a limit before (and I think Univocity is being a little bit over conservative in this case), but as Owen suggests we could probably pretty easily count the number of columns as part of our format guessing and just make sure that the max columns parameter is always set to be big enough.

**Comment by jquartel** [comment]
That would be great!

**Comment by jquartel** [comment]
So I see this appears to be fixed in 3.9.0 which is great, thank you! However, I have loaded one of my new datasets which has 2155 rows and 998 columns and it is very sluggish. Takes sometimes more than a minute to draw the data table to the screen and longer still to apply facets (I've tried both Chrome and Safari on MacOs 15.3.1 on a pretty modern MacBook Pro). What's odd is that I have routinely used datasets with tens of thousands of rows (not as many columns, but definitely dozens) and found the app remarkably agile - only taking multiple seconds when I give it a tricky facet or transform involving a lot of XML parsing. Do we think there may be something amiss with the new implementation that is slowing the app down?

**Comment by thadguidry** [comment]
@jquartel Well, cell data length was allowed to increase, along with the number of columns.  Is it possible that your particular dataset was parsed incorrectly, somehow, or needs to be cleaned up?  My hunch is that now its being allowed to store very large data in some cells, so also check the cell lengths using the Text Length Facet?  Try to narrow down the problem exactly will help us considerably (3.8.7 doesn't exibit the issue as bad? Only some datasets?, After a certain # of columns or value lengths?) and then open a new issue.

**Comment by wetneb** [comment]
If there's a regression in performance between 3.8 and 3.9 on projects with many columns, it's worth opening a new issue about it. It could well be that somehow the column resizing feature has an impact on that (although I intuitively can't think of why).

**Comment by jquartel** [comment]
I can't really check this dataset with earlier versions since 3.9.0 is the only version that will open a CSV with 998 columns. I'm pretty sure there is no issue with the cell length - this type of data rarely has large text cells. i tried cutting out columns which I don't think I need and it's now 640 columns. This seems to have improved drawing the table to screen, but the facets are still quite sluggish. Would you like me to add the CSV to this ticket?

**Comment by wetneb** [comment]
Ah yes, sorry, I didn't realize that the CSV parser issue would also play into it (which is obvious from the context here). Adding the CSV here would be ideal!

**Comment by thadguidry** [comment]
Yes, and @jquartel also note in the new issue if there are problems after importing and seeing within the preview view just before clicking on Create Project button. Or if it's only after clicking on the Create Project button and the main data grid view.

**Comment by jquartel** [comment]
Oh, in fact the datatset may contain commercially sensitive information, so I need to do some further checks to see if I can post it here. There is no problem with the preview view and it loads into the main view quickly enough so I don't think the parsing of the file is the issue. But when I do a straight text facet on a column, the facet pane gets somewhat stuck for a couple of minutes and when attempting to scroll around the main table pane I get large blank areas.

<img width="1018" alt="Image" src="https://github.com/user-attachments/assets/0df946ba-a7bb-4e19-bcd8-f88f4d8f2325" />

**Comment by jquartel** [comment]
With a bit more testing, I think the issue is significantly worse in Safari. For the datasets with with many hundreds of columns (my cell sizes are small, most are blank and none are much more than 100characters), Safari takes a very long time to process any facet action, while Chrome is a little laggy but acceptable. I think, together with the buggy column width behaviour in Safari, this has pushed me to leaving Safari for good. I had been sticking with it because it's browser-level text search behaviour is a little cleaner, but Chrome has improved in this area too so I'm now officially switched.

**Comment by thadguidry** [comment]
Edge browser also works great on Mac and I have friends that use it especially for it's split screen functionality.  Great during reconciling to have a view of the OpenRefine data while clicking links that open in the 2nd split screen (instead of the tiny Suggest preview)
